src/reconcile.py

```diff
@@ -0,0 +1,274 @@
+    )
+
+
+def _local_and_server(
+    path_info: types_.PathInfo, table_row: types_.TableRow, discourse: Discourse
+) -> (types_.UpdateAction | types_.NoopAction | types_.CreateAction | types_.DeleteAction):
+    """Return an update or noop action depending on if the content or navlink title has changed.
+
+    Args:
+        path_info: Information about the local documentation file.
+        table_row: A row from the navigation table.
+        discourse: A client to the documentation server.
+
+    Returns:
+        A page update or noop action.
+
+    Raises:
+        ReconcilliationError: if the table path or level do not match for the path info and table
+            row.
+    """
+    if path_info.level != table_row.level:
+        raise exceptions.ReconcilliationError(
+            f"internal error, level mismatch, {path_info=!r}, {table_row=!r}"
+        )
+    if path_info.table_path != table_row.path:
+        raise exceptions.ReconcilliationError(
+            f"internal error, table path mismatch, {path_info=!r}, {table_row=!r}"
+        )
+
+    # Is a directory locally and a grouping on the server
+    if path_info.local_path.is_dir() and table_row.is_group:
+        if table_row.navlink.title == path_info.navlink_title:
+            return types_.NoopAction(
+                action=types_.Action.NOOP,
+                level=path_info.level,
+                path=path_info.table_path,
+                navlink=table_row.navlink,
+                content=None,
+            )
+        return types_.UpdateAction(
+            action=types_.Action.UPDATE,
+            level=path_info.level,
+            path=path_info.table_path,
+            navlink_change=types_.NavlinkChange(
+                old=table_row.navlink,
+                new=types_.Navlink(title=path_info.navlink_title, link=table_row.navlink.link),
+            ),
+            content_change=types_.ContentChange(old=None, new=None),
+        )
+
+    # Is a directory locally and a page on the server
+    if path_info.local_path.is_dir():
+        # This is an edge case that can't actually occur because table_row.is_group is based on
+        # whether the navlink link is None so this case would have been caught in the local
+        # directory and server group case
+        if table_row.navlink.link is None:  # pragma: no cover
+            raise exceptions.ReconcilliationError(
+                f"internal error, expecting link on table row, {path_info=!r}, {table_row=!r}"
+            )
+        return types_.DeleteAction(
+            action=types_.Action.DELETE,
+            level=path_info.level,
+            path=path_info.table_path,
+            navlink=table_row.navlink,
+            content=discourse.retrieve_topic(url=table_row.navlink.link),
+        )
+
+    # Is a page locally and a grouping on the server
+    if table_row.is_group:
+        return types_.CreateAction(
```

Comment:
The server isn't really aware of groups other than on the navigation table which is always rebuilt based on the local state. This means that just creating the page is enough because the group will be removed from the navigation table automatically as it is not present locally anymore. Does that make sense? I'll add that clarification to the comment as well

Reply:
This has now been pushed to this PR